### PR TITLE
[EOC-406] Terminate other user sessions after password changes

### PR DESCRIPTION
**Pull request by alemikolo**
To show notification regarding no authorization or end of user's session I had to restore error capturing in action's catch statement and pass this error to the notification, and there, by checking its type display a notification with timer feature or not.

I had a problem with a time value to the notification message. Regardless of the way I was passing it, there were no spaces between the time value and the rest of the notification's message. I have no idea why this was happening. I found some workaround and it's working but I don't know if this should be written in that way: File NotificationWithTimer.jsx, line 41-49.

session-terminate

### CODE REVIEW

File: client/src/common/components/Activities/model/actions.js [hunk 1]
```diff
@@ -31,11 +31,12 @@ export const fetchActivities = page => dispatch =>
...
       );
     })
-    .catch(() => {
+    .catch(err => {
       dispatch(fetchActivitiesFailure());
       createNotificationWithTimeout(
         dispatch,
         NotificationType.ERROR,
-        'Failed to fetch activities. Please try again.'
+        { notificationId: 'Failed to fetch activities. Please try again.' },
```
Comment (adamklepacz) [inline client/src/common/components/Activities/model/actions.js]: shouldn't this be an id, not an actual message?

File: client/src/common/constants/variables.js [hunk 2]
```diff
@@ -1,6 +1,7 @@
 export const COMPANY_NAME = 'Sandstream Development Sp. z o.o.';
 export const COMPANY_PAGE_URL = 'https://sandstream.pl/';
 export const NOTIFICATION_TIMEOUT = 5000;
+export const REDIRECT_TIMEOUT = 10000;
```
Comment (adamklepacz) [inline client/src/common/constants/variables.js]: sort alphabetically

File: client/src/locales/en.json [hunk 3]
```diff
@@ -106,6 +106,7 @@
   "user.auth.reset-password.email-label": "Enter email",
   "user.auth.reset-password.button-content": "Reset password",
   "user.auth.pass-recovery-form.heading": "Set new password",
+  "user.auth.session-ended": "Your session has ended, you will be redirected to the login page in {data} s.",
```
Comment (adamklepacz) [inline client/src/locales/en.json]: will it countdown? :D

File: client/src/modules/notification/components/NotificationWithTimer.jsx [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+      time
+    };
+  }
+
+  componentDidMount() {
+    this.createInterval();
+  }
+
+  componentWillUnmount() {
+    clearInterval(this.timer);
```
Comment (adamklepacz) [inline client/src/modules/notification/components/NotificationWithTimer.jsx]: I think this.timer should be declared in the constructor with some default value

File: server/controllers/userAuth.js [hunk 5]
```diff
@@ -411,6 +411,25 @@ const changePassword = (req, res) => {
...
+      const regexp = new RegExp(userId);
+
+      return db
+        .collection('sessions')
+        .find({
+          _id: { $ne: sessionID },
+          session: regexp
+        })
+        .forEach(session => {
+          const { _id: id } = session;
```
Comment (adamklepacz) [inline server/controllers/userAuth.js]: line of a break

Comment (adamklepacz) [inline server/controllers/userAuth.js]: do you need to map it from _id to id?

File: server/controllers/userAuth.js [hunk 6]
```diff
@@ -411,6 +411,26 @@ const changePassword = (req, res) => {
...
+      } = req;
+      const regexp = new RegExp(userId);
+
+      return db
+        .collection('sessions')
+        .find({
+          _id: { $ne: sessionID },
+          session: regexp
+        })
+        .forEach(session => {
```
Comment (marekrozmus) [inline server/controllers/userAuth.js]: `({ _id }) => store.destroy(_id)`

File: client/src/modules/notification/index.jsx [hunk 7]
```diff
@@ -2,36 +2,37 @@ import React, { Fragment } from 'react';
...
               return (
                 <li className="notification__list-item" key={id}>
-                  <MessageBox type={item.type}>
-                    <FormattedMessage
+                  {redirect ? (
+                    <NotificationWithTimer
                       id={notificationId}
-                      values={{ data: <strong>{itemName}</strong> }}
+                      redirect={redirect}
+                      time={REDIRECT_TIMEOUT / 1000}
```
Comment (marekrozmus) [inline client/src/modules/notification/index.jsx]: `timeout`

File: client/src/modules/notification/model/actions.js [hunk 8]
```diff
@@ -16,12 +20,21 @@ const removeNotification = payload => ({
 export const createNotificationWithTimeout = (
   dispatch,
   type,
-  notification,
+  message,
+  err = null,
   timeout = NOTIFICATION_TIMEOUT
 ) => {
+  const forbiddenError = err instanceof ForbiddenException;
+
```
Comment (adamklepacz) [inline client/src/modules/notification/model/actions.js]: remove thath empty line

File: client/src/modules/notification/model/actions.js [hunk 9]
```diff
@@ -16,12 +20,21 @@ const removeNotification = payload => ({
...
   const id = uniqueId('notification_');
-  dispatch(addNotification({ id, type, notification }));
+  const delay = forbiddenError ? REDIRECT_TIMEOUT : timeout;
+  const notification = forbiddenError
+    ? { notificationId: 'user.auth.session-ended' }
```
Comment (adamklepacz) [inline client/src/modules/notification/model/actions.js]: what if we gonna have other notifications with counter??
wouldn't be better to pass notifictaionID as parameter, as you do in regular notification via message parameter ??

File: server/controllers/userAuth.js [hunk 10]
```diff
@@ -411,6 +411,22 @@ const changePassword = (req, res) => {
...
+        sessionStore: store,
+        sessionStore: { db }
+      } = req;
+      const regexp = new RegExp(userId);
+
+      return db
+        .collection('sessions')
+        .find({
+          _id: { $ne: sessionID },
+          session: regexp
```
Comment (adamklepacz) [inline server/controllers/userAuth.js]: will this query find userId in his session ?? wouldn't it destroy all the existing sessions in the database, even those ones that not belong to him??

cause session prop in DB looks like this 
`{"cookie":{"originalMaxAge":null,"expires":null,"httpOnly":true,"path":"/"},"passport":{"user":"5c9dc5a619ae7924074940f4"}}`

File: client/src/modules/notification/model/actions.js (hunk 9 above)
Comment (alemikolo) [reply client/src/modules/notification/model/actions.js]: If we gonna have other notifications with I think we pass notificationId or some other prop from specific actions to the notification dispatcher... In `createNotificationWithTimeout` we will be able to decide what kind of notification should we display.
But in this case, in all actions in the app, I would have to check the type of the error and pass right notifactionId or dispatch different action. I didn't want to do all these things in all app's actions. So I decided to pass only error from catch to notification `createNotificationWithTimeout` and all the logic write in one place.

File: server/controllers/userAuth.js (hunk 10 above)
Comment (alemikolo) [reply server/controllers/userAuth.js]: We create a regexp with userId in line 420: `const regexp = new RegExp(userId)`. Mongo recognizes regexp and uses it to find this user id in the session's string. Other user's session remain.

**Pull request merged**